File: programmers/여행경로.py
import copy
ans = []
def solution(tickets):
    answer = ["ICN"]
    airport = "ICN"
    dic = {}
    
    for ticket in tickets :
        dep = ticket[0]
        des = ticket[1]
        if dep not in dic :
            dic[dep] = [des]
        else :
            dic[dep].append(des)
    for d in dic:
        dic[d].sort()
    
    def dfs(city,count) :
        
        global ans
        
        if count == len(tickets) :
            temp = copy.deepcopy(answer)
            ans.append(temp)
            return
        if city not in dic : # 여기땜에 자꾸 오류났었음.. dict의 경우 key에 해당하는 value 처리를 해줬어야함.. [["ICN", "A"], ["ICN", "B"], ["B", "ICN"]] 인 경우 A : key error
            return
        city_list = dic[city]

        if len(city_list) == 0:
            return
        for airport in city_list :
            
            dic[city].remove(airport)
            answer.append(airport)
            dfs(airport, count+1)
            answer.pop()
            dic[city].append(airport)
            dic[city].sort()
    dfs("ICN",0)
    
    return ans[0]

File: programmers/코테대비(2022.01.09)/여행경로.java
import java.util.*;
class Solution {
    static Map<String,Object> map;
    static ArrayList<String> answer;
    static int length = 0;
    static String[] ans = null;
    static void dfs(String city,int count){
        
        if(count == length && ans == null){
            ans = answer.toArray(new String[answer.size()]);
            return;
        } 

        if(!map.containsKey(city)){
            return;
        }
        
        List<String> arr = (ArrayList)map.get(city);
        
        if(arr.size()==0){
            return;
        }
        
        for(String airport : arr){
            
            List<String> list1 = (ArrayList)map.get(city);
            
            List<String> temp = new ArrayList();
            for(String str : list1){
                if(!str.equals(airport)){
                    temp.add(str);
                }
            }
            map.put(city,temp);
            answer.add(airport);
            
            dfs(airport,count+1);
            
            answer.remove(answer.size()-1);
            
            List<String> list2 = (ArrayList)map.get(city);
            temp = new ArrayList();
            for(String str : list2){

                temp.add(str);
            }
            temp.add(airport);
            Collections.sort(temp);
            map.put(city,temp);
            
        }
        
    }
    
    public String[] solution(String[][] tickets) {

        map = new HashMap();
        length = tickets.length;
        answer = new ArrayList();
        for(String[] str : tickets){
            String start = str[0];
            String end = str[1];
            
            if(map.containsKey(start)){
                List<String> arr = (ArrayList)map.get(start);
                arr.add(end);
                Collections.sort(arr);
                map.put(start,arr);
            }else{
                List<String> arr = new ArrayList();
                arr.add(end);
                map.put(start, arr);
            }
        }
        answer.add("ICN");
        dfs("ICN",0);
        return ans;
    }
}
